scvi/inference/jvae_trainer.py

```diff
@@ -0,0 +1,386 @@
+        if verbose:
+            logger.info("ELBO : %.4f" % elbo)
+        return elbo
+
+
+class JVAETrainer(Trainer):
+    r"""
+    The VariationalInference class for the unsupervised training of an autoencoder.
+
+    :param model: A model instance from class ``JVAE``
+    :param discriminator: A model instance of a classifier (with logit output)
+    :param gene_dataset_list: list of gene_dataset instance like ``[CortexDataset(), SmfishDataset()]``
+    :param train_size: Train size on cells
+    :param kappa: float to weight the discriminator loss
+    :param n_epochs_kl_warmup: Number of epochs for linear warmup of KL(q(z|x)||p(z)) term. After `n_epochs_kl_warmup`,
+        the training objective is the ELBO. This might be used to prevent inactivity of latent units, and/or to
+        improve clustering of latent space, as a long warmup turns the model into something more of an autoencoder.
+    :param kwargs: Other keywords arguments from the general Trainer class.
+    """
+
+    default_metrics_to_monitor = ["elbo"]
+
+    def __init__(
+        self,
+        model: nn.Module,
+        discriminator: nn.Module,
+        gene_dataset_list: List[GeneExpressionDataset],
+        train_size: float = 0.8,
+        use_cuda: bool = True,
+        kappa: float = 1.0,
+        n_epochs_kl_warmup: int = 400,
+        **kwargs
+    ):
+
+        super().__init__(model, gene_dataset_list[0], use_cuda=use_cuda, **kwargs)
+        self.n_epochs_kl_warmup = n_epochs_kl_warmup
+        self.kappa_target = kappa
+        self.all_dataset = [
+            self.create_posterior(
+                gene_dataset=gd, type_class=partial(JPosterior, mode=i)
+            )
+            for i, gd in enumerate(gene_dataset_list)
+        ]
+        self.n_dataset = len(self.all_dataset)
+        self.all_train, self.all_test = list(
+            zip(
+                *[
+                    self.train_test(
+                        model, gd, train_size, type_class=partial(JPosterior, mode=i)
+                    )
+                    for i, gd in enumerate(gene_dataset_list)
+                ]
+            )
+        )
+        for i, d in enumerate(self.all_train):
+            self.register_posterior("train_%d" % i, d)
+            d.to_monitor = ["elbo"]
+
+        for i, d in enumerate(self.all_test):
+            self.register_posterior("test_%d" % i, d)
+            d.to_monitor = ["elbo"]
+
+        self.discriminator = discriminator
+        if self.use_cuda:
+            self.discriminator.cuda()
+
+        self.kl_weight = None
+        self.kappa = None
+        self.compute_metrics_time = None
+        self.n_epochs = None
+
+        self.track_disc = []
+
+    def on_epoch_begin(self):
+        if self.n_epochs_kl_warmup is not None:
+            self.kl_weight = min(1, self.epoch / self.n_epochs_kl_warmup)
+        else:
+            self.kl_weight = 1.0
+        self.kappa = (self.epoch / self.n_epochs) * self.kappa_target
```

Comment:
I did not remember you did that in the manuscript. what is self.n_epochs, do we have a docstring somehwere ?

Reply:
Oh good one! I tried this recently and left it. I think I can remove this warmup which didn't make much a difference.

Reply:
ok perfect